### ISSUE TITLE
requirements.txt is missing

### ISSUE THREAD

**Issue by ikuyamada** [facebookresearch/NPM#1]
Hi!
Thank you for sharing your amazing work!
I am now working on reproducing your experiments, but it seems that `requirements.txt` is currently missing in this repository.
I am sorry if you are still working on this.

**Comment by SongHoSung** [comment]
Hi~ 
I'm just a watcher, but I'll share what I've tested after receiving the code.
I'm running some closed-set evaluation, but so far it's going well.

conda install -c conda-forge faiss-gpu (or cpu version)
install fairscale : https://github.com/facebookresearch/fairscale
pip install torch, numpy datasets scipy transformers hydra hydra-core pytorch_lighting==1.7 pathos joblib
(torch_lighting version 1.8 requires code modification.)

#### If GLIBCXX_3.4.26' not found error occurs
sudo add-apt-repository ppa:ubuntu-toolchain-r/test
sudo apt-get update
sudo apt install wget gcc-8 unzip libssl1.0.0 software-properties-common
sudo apt-get upgrade libstdc++6 (if not resolved, use the command below)
sudo apt-get install --only-upgrade libstdc++6

I want to try training, but it seems that preprocessing and some codes for training are missing. If the author sees this article, please upload additional code for training.

**Comment by shmsw25** [comment]
Hi @ikuyamada @SongHoSung. Thanks for reporting the issue. Turns out I did not update `.gitignore` when adding files.. 😂 😂  The repo has been updated with new files, and it should have all codes necessary for inference and training. Please kindly let us know if you run into any other issues (we're still in the process of testing the code).

> If GLIBCXX_3.4.26' not found error occurs

Is this an error? I do not think I have encountered an error message like this... If this is an error, could you clarify which package is causing this error while being installed?

**Comment by SongHoSung** [comment]
@shmsw25 
When the environment was reconstructed with the shared requirement.txt, the error did not occur.
It is not possible to know which package caused the error because the error log was not left, but it seems that the latest version of some package requires the corresponding library.

And it seems that the file below has not been uploaded yet.
'data/fuzzy_verbalizers/agn.txt'

**Comment by shmsw25** [comment]
Interesting, not exactly sure why this problem occurs, but looks like you managed to install all requirements, so I'll hold it unless somebody else is blocked by the same problem.

RE `data/fuzzy_verbalizers/agn.txt`:
This is supposed to be downloaded at [this step](https://github.com/facebookresearch/NPM#download-data) through `scripts/download_data.sh` -- in particular, [these lines](https://github.com/facebookresearch/NPM/blob/main/scripts/download_data.sh#L35-L41). Let me know if you still have the problem after running `scripts/download_data.sh`!

**Comment by SongHoSung** [comment]
Thank you for your kind reply.

**Comment by ikuyamada** [comment]
@shmsw25 @SongHoSung
Thanks so much for your reply and useful info!
Again, congrats on this amazing work!

**Comment by SongHoSung** [comment]
Evaluation seems to take longer than expected, but it worked normally.

100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 3000/3000 [3:09:39<00:00,  3.79s/it]
Task: agn       Accuracy=73.5%

**Comment by shmsw25** [comment]
@SongHoSung Thanks for confirming! 3 hours for the AGNews seems definitely longer than expected... When tested on AGN with `--corpus_data enwiki-0` (instead of `--corpus_data enwiki-0+cc_news`, which will take longer but not more than double), it takes about 6.5min for me to run on 3,000 examples.

Did you place `save` (embeddings and a faiss index) in your local directory, not a remote disk? If you are using a remote disk, this will significantly hurt inference speed.

The other possibility is that your RAM usage is full so you are using the swp space. However, it seems unlikely, given that `enwiki-0+cc_news` does not require that much memory (around 10GB).

Happy to go through in more detail on what's slowing down the inference speed!

**Comment by SongHoSung** [comment]
@shmsw25 
Embeddings and a faiss index are stored locally.
Loading time was less than 300 seconds.

Start loading the embed with (59727420, 1024)...
memory use: 4.6GB
Starting loading save/npm/dstore/enwiki-0/embeddings.faiss_index_IP
load_restricted: None
memory use: 8.9GB
100%|██| 249494/249494 [01:57<00:00, 2119.24it/s]
Finished reading 56.290M tokens from corpus/cc_news/text.npy
memory use: 11.4GB
memory use: 10.9GB
Start loading the embed with (56290444, 1024)...
memory use: 10.9GB
Starting loading save/npm/dstore/cc_news/embeddings.faiss_index_IP
Finish loading the datastore (283sec)

The gpu I used is one A100 40G,
When I checked the memory, there was a margin of more than 1 tera out of the total 1.3T.
I'll let you know later when I find out which part became the bottleneck.

Other tasks took the same amount of time,
When I have free time, I will test it later in another device environment.

100% | 3000/3000 [2:48:25<00:00,  3.37s/it]
Task: yahoo     Accuracy=54.1%

100%| 277/277 [15:32<00:00,  3.37s/it]
Task: rte       Accuracy=53.1%

thank you

**Comment by shmsw25** [comment]
It actually looks like using a union of corpora is causing a significant delay. I just made a quick change and pushed a commit. Could you try the latest version when you get a chance? Hopefully this should significantly improve inference speed. 

Thanks for reporting the issue with the inference speed! I did not realize about this issue before.

**Comment by SongHoSung** [comment]
Fixed the issue with the corrected code~ 
It takes about 10 minutes.

3000/3000 [10:59<00:00,  4.55it/s]
Task: agn       Accuracy=73.5%
